packages/sdk/src/oracle/services/statemachines/disputePrice.ts

```diff
@@ -0,0 +1,46 @@
+import assert from "assert";
+import { Update } from "../update";
+import Store from "../../store";
+import { Signer } from "../../types/ethers";
+import { Handlers as GenericHandlers } from "../../types/statemachine";
+import { InputRequest } from "../../types/state";
+import { ContextClient } from "./utils";
+
+export type Params = InputRequest & { signer: Signer; confirmations: number; currency: string; account: string };
+
+export type Memory = { hash?: string };
+
+export function initMemory(): Memory {
+  return {};
+}
+
+export function Handlers(store: Store): GenericHandlers<Params, Memory> {
+  const update = new Update(store);
+  return {
+    async start(params: Params, memory: Memory) {
+      const { requester, identifier, timestamp, ancillaryData, chainId, signer } = params;
+      assert(chainId === (await signer.getChainId()), "Signer on wrong chainid");
+
+      const oracle = store.read().oracleService(chainId);
+      const tx = await oracle.disputePrice(signer, requester, identifier, timestamp, ancillaryData);
+      memory.hash = tx.hash;
+      return "confirm";
+    },
+    async confirm(params: Params, memory: Memory, context: ContextClient) {
```

Comment:
This is another spot where inheritance might be able to improve things. This confirm function looks identical to the other confirm function. In a class-based handler world, this could be implemented in a base class so it wouldn't need to be copy-pasted.

In fact, this file seems to be copy-pasted outside of the disputePrice call above, so I think much of this code could be eliminated in an inheritance structure.

Reply:
good call, yea there is a lot of code dupe here. when i try to turn these into classes i end up getting stuck with typescript because of the way i index into class methods with a string in the statemachine. It requires more extensive changes to get that working which i am not sure how to do at the moment.  

In more sophisticated infrastructure, the way i've solved this code dupe problem in the past, rather than using inheritance, we can compose state machines, so that you have a transaction waiter, which can be called and resolved by other state machines.  This feature is really useful, so then you can treate these state machines almost as functions themselves, but the code to do this is challenging.